### DISCUSSION TITLE
Global event scripts (for all files)

### DISCUSSION THREAD

**Discussion by klutzer** [AnWeber/httpyac#461] [General]
Is it possible to have a single file containing event scripts (like pre or post-request ones) and make it available for all other .http files in different folders of the project? I'm trying to create a "common.http" file containing global variables and scripts but even importing it to every file, the global event scripts aren't executed for requests outside the file. Any clues about that?

**Comment by AnWeber** [comment]
@klutzer It is already possible, but in my eyes not the right approach. `# @ref` is executed exactly where it is contained in the region and can therefore be used.

> name.http
```
{{
  console.info("prerequest script");
}}
###
# @name after
{{
  console.info("response script");
}}
```
> import.http
```
# @import ./name.http
POST https://httpbin.org/anything

{{
  console.info("request");
}}
# @ref after
```

The much better solution I use is to call the [plugin api in a `.httpyac.js`](https://httpyac.github.io/config/#configurehooks). There you can simply enter a hook for [request](https://httpyac.github.io/plugins/plugin-api.html#onrequest-hook) and [response](https://httpyac.github.io/plugins/plugin-api.html#onresponse-hook) and they are always executed. Either this file is located in the root of your project or you configure the location using the global variable `HTTPYAC_PLUGIN`.

> .httpyac.js

```js
module.exports = {
	configureHooks: function (api) {
		 api.hooks.onRequest.addHook('getCoffee', function (request) {
                    request.headers = Object.assign({
                      'X-Coffee':  'coffee with milk, no sugar',
                     }, request.headers);
                  });
              api.hooks.onResponse.addHook('getCoffee', function (response) {
                  response.body = 'coffee with milk, no sugar';
                  delete response.parsedBody;
                  delete response.prettyPrintBody;
                });
	}
}
```

The documentation is relatively difficult to understand, I'm just realizing. If you have any questions, just let me know. The best way to develop such a plugin is with the CLI. NodeJS may cache the script and that makes creating and executing the script in VSCode difficult.

**Reply by klutzer** [reply]
Thank you for replying this thread.

Can I use the second approach to get and set global variables (for later use in different files)?

The `@ref` approach isn't what I need because I have to reference in each endpoint for every file, so I'll probably take with your second suggestion.

**Comment by AnWeber** [comment]
For variables, I would have created a separate [hook `provideVariablesHook`](https://httpyac.github.io/plugins/plugin-api.html#providevariableshook) that is always called once before the regions are executed. This is intended for requesting variables from a central vault, for example.
Of course this also works in the other hooks, if this makes more sense for your use case (see [variableparser](https://github.com/AnWeber/httpyac/blob/main/src/plugins/core/parse/variableHttpRegionParser.ts#L22)). 

```js
module.exports = {
	configureHooks: function (api) {
		 api.hooks.onRequest.addHook('getCoffee', function (request, context) {
                    utils.setVariableInContext(
                        {
                          foo: "bar",
                        },
                        context
                      );
                  });
	}
}
```

**Reply by klutzer** [reply]
In this case you've set variable using utils (requiring from httpyac), right?

Inside the hook I'll have to call an authentication request, as the already existing oauth feature didn't work for me because the provider uses a special grant_type and I also have to authenticate sending username and password along with clientId and secret, so I'm trying to automate the authentication process using those hooks. Is there a way to conditionally call a request from specific http file in the hook? Or do you suggest another approach?

**Reply by AnWeber** [reply]
I don't think the way makes sense, although it probably could work. I find a direct call with a fetch like API easier. For me, that's why I built in a helper method ([httpClientProvider.exchange](https://github.com/AnWeber/httpyac/blob/main/src/models/httpClientProvider.ts#L27) that I use for such cases. See my oauth implementation how to use it.
- https://github.com/AnWeber/httpyac/blob/main/src/plugins/oauth2/flow/requestOpenIdInformation.ts#L36
- [oauth2VariableReplacer](https://github.com/AnWeber/httpyac/blob/main/src/plugins/oauth2/oauth2VariableReplacer.ts#L9) - implemetation with Token Caching and update AccessToken using RefreshToken
- [basicAuthVariableReplacer](https://github.com/AnWeber/httpyac/blob/main/src/plugins/http/basicAuthVariableReplacer.ts#L5) simple auth approach
